pkg/logs/input/journald/launcher.go

```diff
@@ -0,0 +1,95 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2018 Datadog, Inc.
+
+package journald
+
+import (
+	log "github.com/cihub/seelog"
+
+	"github.com/DataDog/datadog-agent/pkg/logs/auditor"
+	"github.com/DataDog/datadog-agent/pkg/logs/config"
+	"github.com/DataDog/datadog-agent/pkg/logs/pipeline"
+	"github.com/DataDog/datadog-agent/pkg/logs/restart"
+)
+
+// ErrorHandler handles tailing errors
+type ErrorHandler interface {
+	Handle(error TailError)
+}
+
+// Launcher is in charge of starting and stopping new journald tailers
+type Launcher struct {
+	sources          []*config.LogSource
+	pipelineProvider pipeline.Provider
+	auditor          *auditor.Auditor
+	tailers          map[string]*Tailer
+}
+
+// New returns a new Launcher.
+func New(sources []*config.LogSource, pipelineProvider pipeline.Provider, auditor *auditor.Auditor) *Launcher {
+	journaldSources := []*config.LogSource{}
+	for _, source := range sources {
+		if source.Config.Type == config.JournaldType {
+			journaldSources = append(journaldSources, source)
+		}
+	}
+	return &Launcher{
+		sources:          journaldSources,
+		pipelineProvider: pipelineProvider,
+		auditor:          auditor,
+		tailers:          make(map[string]*Tailer),
+	}
+}
+
+// Start starts new tailers.
+func (l *Launcher) Start() {
+	for _, source := range l.sources {
+		identifier := source.Config.Path
+		if _, exists := l.tailers[identifier]; exists {
+			// set up only one tailer per journal
+			continue
+		}
+		tailer, err := l.setupTailer(source)
+		if err != nil {
+			log.Warn("Could not set up journald tailer: ", err)
+		} else {
+			l.tailers[identifier] = tailer
+		}
+	}
+}
+
+// Stop stops all active tailers
+func (l *Launcher) Stop() {
+	stopper := restart.NewParallelStopper()
+	for identifier, tailer := range l.tailers {
+		stopper.Add(tailer)
+		delete(l.tailers, identifier)
+	}
+	stopper.Stop()
+}
+
+// Handle keeps all tailers alive, restarting them when a fatal error occurs.
+func (l *Launcher) Handle(tailError TailError) {
+	log.Error(tailError.err)
```

Comment:
do you have example of errors that you saw happening?

Reply:
I did not encounter this use case yet, we just want to avoid stop collecting logs when such an error occur.

Reply:
Hum, I think this is over engineered for now. Logging the error is enough, and when this happens we will have more context and will be able to chose the appropriate behavior.
With the current implementation, we stop and restart the tailer, and if the error persists, we would be doing this in a loop, which wouldn't be great performance wise.
I'd much rather have the agent stop collecting systemd logs when it encounters an error, and people share feedback on what went wrong, and us fixing it / catching it properly